File: autocontext/src/autocontext/session/coordinator.py
"""Coordinator-first execution for multi-worker missions (AC-515).

Domain concepts:
- Worker: entity tracking one delegated unit of work with lineage
- WorkerStatus: lifecycle (pending → running → completed/failed/redirected)
- Coordinator: aggregate root owning plan, workers, fan-out/fan-in, events
- CoordinatorEvent: structured event for observability
"""

from __future__ import annotations

import uuid
from datetime import UTC, datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


def _now() -> str:
    return datetime.now(UTC).isoformat()


# ---- Value Objects ----


class WorkerStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    REDIRECTED = "redirected"


class CoordinatorEventType(StrEnum):
    COORDINATOR_CREATED = "coordinator_created"
    WORKER_DELEGATED = "worker_delegated"
    WORKER_STARTED = "worker_started"
    WORKER_COMPLETED = "worker_completed"
    WORKER_FAILED = "worker_failed"
    WORKER_REDIRECTED = "worker_redirected"
    FAN_OUT = "fan_out"
    FAN_IN = "fan_in"


_ACTIVE_STATUSES = frozenset({WorkerStatus.PENDING, WorkerStatus.RUNNING})
_RETRYABLE_STATUSES = frozenset({WorkerStatus.FAILED, WorkerStatus.REDIRECTED})


class CoordinatorEvent(BaseModel):
    """Immutable event in the coordinator event stream."""

    event_id: str = Field(default_factory=lambda: uuid.uuid4().hex[:12])
    event_type: CoordinatorEventType
    timestamp: str = Field(default_factory=_now)
    payload: dict[str, Any] = Field(default_factory=dict)

    model_config = {"frozen": True}


# ---- Entities ----


class Worker(BaseModel):
    """Entity tracking one delegated unit of work.

    Workers have lineage: a retried worker references its parent.
    """

    worker_id: str = Field(default_factory=lambda: uuid.uuid4().hex[:12])
    task: str
    role: str
    status: WorkerStatus = WorkerStatus.PENDING
    result: str = ""
    error: str = ""
    redirect_reason: str = ""
    parent_worker_id: str = ""
    created_at: str = Field(default_factory=_now)
    completed_at: str = ""

    @classmethod
    def create(
        cls,
        task: str,
        role: str,
        parent_worker_id: str = "",
    ) -> Worker:
        return cls(task=task, role=role, parent_worker_id=parent_worker_id)

    def start(self) -> None:
        self._require_status({WorkerStatus.PENDING}, action="start worker")
        self.status = WorkerStatus.RUNNING

    def complete(self, result: str) -> None:
        self._require_status({WorkerStatus.RUNNING}, action="complete worker")
        self.status = WorkerStatus.COMPLETED
        self.result = result
        self.completed_at = _now()

    def fail(self, error: str = "") -> None:
        self._require_status({WorkerStatus.RUNNING}, action="fail worker")
        self.status = WorkerStatus.FAILED
        self.error = error
        self.completed_at = _now()

    def redirect(self, new_task: str = "", reason: str = "") -> None:
        self._require_status({WorkerStatus.RUNNING}, action="redirect worker")
        self.status = WorkerStatus.REDIRECTED
        self.redirect_reason = reason
        self.completed_at = _now()

    @property
    def is_active(self) -> bool:
        return self.status in _ACTIVE_STATUSES

    def _require_status(
        self,
        allowed: set[WorkerStatus] | frozenset[WorkerStatus],
        action: str,
    ) -> None:
        if self.status not in allowed:
            msg = f"Cannot {action} from status={self.status}"
            raise ValueError(msg)


# ---- Aggregate Root ----


class Coordinator(BaseModel):
    """Aggregate root: owns plan, workers, and fan-out/fan-in.

    Create via Coordinator.create().
    """

    coordinator_id: str = Field(default_factory=lambda: uuid.uuid4().hex[:12])
    session_id: str
    goal: str
    workers: list[Worker] = Field(default_factory=list)
    events: list[CoordinatorEvent] = Field(default_factory=list)
    created_at: str = Field(default_factory=_now)

    @classmethod
    def create(cls, session_id: str, goal: str) -> Coordinator:
        coord = cls(session_id=session_id, goal=goal)
        coord._emit(CoordinatorEventType.COORDINATOR_CREATED, {"goal": goal})
        return coord

    # -- Worker management --

    def delegate(self, task: str, role: str, parent_worker_id: str = "") -> Worker:
        """Create and register a new worker."""
        worker = Worker.create(task=task, role=role, parent_worker_id=parent_worker_id)
        self.workers.append(worker)
        self._emit(CoordinatorEventType.WORKER_DELEGATED, {
            "worker_id": worker.worker_id,
            "task": task,
            "role": role,
        })
        return worker

    def fan_out(self, tasks: list[dict[str, str]]) -> list[Worker]:
        """Delegate multiple independent tasks at once."""
        workers = [self.delegate(**t) for t in tasks]
        self._emit(CoordinatorEventType.FAN_OUT, {
            "worker_ids": [w.worker_id for w in workers],
            "count": len(workers),
        })
        return workers

    def fan_in(self) -> list[str]:
        """Collect results from all completed workers."""
        results = [w.result for w in self.workers if w.status == WorkerStatus.COMPLETED]
        self._emit(CoordinatorEventType.FAN_IN, {
            "result_count": len(results),
        })
        return results

    def complete_worker(self, worker_id: str, result: str) -> None:
        """Mark a worker as completed with its result."""
        worker = self._get_worker(worker_id)
        worker.complete(result=result)
        self._emit(CoordinatorEventType.WORKER_COMPLETED, {
            "worker_id": worker_id,
        })

    def stop_worker(self, worker_id: str, reason: str = "") -> None:
        """Redirect a worker away from its current task."""
        worker = self._get_worker(worker_id)
        worker.redirect(reason=reason)
        self._emit(CoordinatorEventType.WORKER_REDIRECTED, {
            "worker_id": worker_id,
            "reason": reason,
        })

    def retry(self, worker_id: str, new_task: str = "") -> Worker:
        """Create a continuation worker linked to a failed/redirected one."""
        parent = self._get_worker(worker_id)
        if parent.status not in _RETRYABLE_STATUSES:
            msg = (
                "Cannot retry worker unless it is failed or redirected "
                f"(status={parent.status})"
            )
            raise ValueError(msg)
        task = new_task or parent.task
        return self.delegate(task=task, role=parent.role, parent_worker_id=parent.worker_id)

    # -- Queries --

    @property
    def active_workers(self) -> list[Worker]:
        return [w for w in self.workers if w.is_active]

    @property
    def completed_workers(self) -> list[Worker]:
        return [w for w in self.workers if w.status == WorkerStatus.COMPLETED]

    # -- Internal --

    def _get_worker(self, worker_id: str) -> Worker:
        for w in self.workers:
            if w.worker_id == worker_id:
                return w
        msg = f"Worker {worker_id} not found"
        raise KeyError(msg)

    def _emit(self, event_type: CoordinatorEventType, payload: dict[str, Any]) -> None:
        self.events.append(CoordinatorEvent(
            event_type=event_type,
            payload={"coordinator_id": self.coordinator_id, **payload},
        ))

File: ts/src/session/coordinator.ts
/**
 * Coordinator-first multi-worker execution (AC-515 TS parity).
 *
 * Port of Python autocontext.session.coordinator.
 */

import { randomUUID } from "node:crypto";

export const WorkerStatus = {
  PENDING: "pending",
  RUNNING: "running",
  COMPLETED: "completed",
  FAILED: "failed",
  REDIRECTED: "redirected",
} as const;
export type WorkerStatus = (typeof WorkerStatus)[keyof typeof WorkerStatus];

export const CoordinatorEventType = {
  COORDINATOR_CREATED: "coordinator_created",
  WORKER_DELEGATED: "worker_delegated",
  WORKER_STARTED: "worker_started",
  WORKER_COMPLETED: "worker_completed",
  WORKER_FAILED: "worker_failed",
  WORKER_REDIRECTED: "worker_redirected",
  FAN_OUT: "fan_out",
  FAN_IN: "fan_in",
} as const;
export type CoordinatorEventType = (typeof CoordinatorEventType)[keyof typeof CoordinatorEventType];

const ACTIVE_STATUSES = new Set<WorkerStatus>([WorkerStatus.PENDING, WorkerStatus.RUNNING]);
const RETRYABLE_STATUSES = new Set<WorkerStatus>([WorkerStatus.FAILED, WorkerStatus.REDIRECTED]);

export interface CoordinatorEvent {
  readonly eventId: string;
  readonly eventType: CoordinatorEventType;
  readonly timestamp: string;
  readonly payload: Record<string, unknown>;
}

export class Worker {
  readonly workerId: string;
  readonly task: string;
  readonly role: string;
  status: WorkerStatus = WorkerStatus.PENDING;
  result: string = "";
  error: string = "";
  redirectReason: string = "";
  readonly parentWorkerId: string;

  private constructor(opts: { task: string; role: string; parentWorkerId?: string }) {
    this.workerId = randomUUID().slice(0, 12);
    this.task = opts.task;
    this.role = opts.role;
    this.parentWorkerId = opts.parentWorkerId ?? "";
  }

  static create(opts: { task: string; role: string; parentWorkerId?: string }): Worker {
    return new Worker(opts);
  }

  start(): void {
    this.requireStatus(new Set([WorkerStatus.PENDING]), "start worker");
    this.status = WorkerStatus.RUNNING;
  }

  complete(result: string): void {
    this.requireStatus(new Set([WorkerStatus.RUNNING]), "complete worker");
    this.status = WorkerStatus.COMPLETED;
    this.result = result;
  }

  fail(error: string = ""): void {
    this.requireStatus(new Set([WorkerStatus.RUNNING]), "fail worker");
    this.status = WorkerStatus.FAILED;
    this.error = error;
  }

  redirect(reason: string = ""): void {
    this.requireStatus(new Set([WorkerStatus.RUNNING]), "redirect worker");
    this.status = WorkerStatus.REDIRECTED;
    this.redirectReason = reason;
  }

  get isActive(): boolean { return ACTIVE_STATUSES.has(this.status); }

  private requireStatus(allowed: Set<WorkerStatus>, action: string): void {
    if (!allowed.has(this.status)) {
      throw new Error(`Cannot ${action} from status=${this.status}`);
    }
  }
}

export class Coordinator {
  readonly coordinatorId: string;
  readonly sessionId: string;
  readonly goal: string;
  readonly workers: Worker[] = [];
  readonly events: CoordinatorEvent[] = [];

  private constructor(sessionId: string, goal: string) {
    this.coordinatorId = randomUUID().slice(0, 12);
    this.sessionId = sessionId;
    this.goal = goal;
  }

  static create(sessionId: string, goal: string): Coordinator {
    const coord = new Coordinator(sessionId, goal);
    coord.emit(CoordinatorEventType.COORDINATOR_CREATED, { goal });
    return coord;
  }

  delegate(task: string, role: string, parentWorkerId?: string): Worker {
    const worker = Worker.create({ task, role, parentWorkerId });
    this.workers.push(worker);
    this.emit(CoordinatorEventType.WORKER_DELEGATED, { workerId: worker.workerId, task, role });
    return worker;
  }

  fanOut(tasks: Array<{ task: string; role: string }>): Worker[] {
    const workers = tasks.map((t) => this.delegate(t.task, t.role));
    this.emit(CoordinatorEventType.FAN_OUT, { count: workers.length });
    return workers;
  }

  fanIn(): string[] {
    const results = this.workers
      .filter((w) => w.status === WorkerStatus.COMPLETED)
      .map((w) => w.result);
    this.emit(CoordinatorEventType.FAN_IN, { resultCount: results.length });
    return results;
  }

  completeWorker(workerId: string, result: string): void {
    this.getWorker(workerId).complete(result);
    this.emit(CoordinatorEventType.WORKER_COMPLETED, { workerId });
  }

  stopWorker(workerId: string, reason: string = ""): void {
    this.getWorker(workerId).redirect(reason);
    this.emit(CoordinatorEventType.WORKER_REDIRECTED, { workerId, reason });
  }

  retry(workerId: string, newTask?: string): Worker {
    const parent = this.getWorker(workerId);
    if (!RETRYABLE_STATUSES.has(parent.status)) {
      throw new Error(
        `Cannot retry worker unless it is failed or redirected (status=${parent.status})`,
      );
    }
    return this.delegate(newTask ?? parent.task, parent.role, parent.workerId);
  }

  get activeWorkers(): Worker[] {
    return this.workers.filter((w) => w.isActive);
  }

  private getWorker(workerId: string): Worker {
    const w = this.workers.find((w) => w.workerId === workerId);
    if (!w) throw new Error(`Worker ${workerId} not found`);
    return w;
  }

  private emit(eventType: CoordinatorEventType, payload: Record<string, unknown>): void {
    this.events.push({
      eventId: randomUUID().slice(0, 12),
      eventType,
      timestamp: new Date().toISOString(),
      payload: { coordinatorId: this.coordinatorId, ...payload },
    });
  }
}
